### ISSUE TITLE
Changes in experiment state on Stage are not picked up by Fenix after two restarts

### ISSUE THREAD

**Issue by data-sync-user** [mozilla/experimenter#6939]
**[Affected Versions]:**

* Firefox Nightly 97.0a1 (AS: 86.2.1) (Build #2015850675, 3058f5257+) (AC: 97.0.20211208143435, 8299d94751) (GV: 97.0a1-20211207040938)

**[Affected Platforms]:**

* Android 9 (Samsung Galaxy S8)

**[Prerequisites]:**

* Have a Firefox for Android experiment live on stage.

**[Steps to reproduce]:**
1. Create a Signed APK from a branch of the fenix project set to Stage (e.g. Firefox Nightly from the “main” branch created based on [this document](https://docs.google.com/document/d/1WqG9YnVnMJ3hk7vcym-kmxT2bsx92HqDagtl3nMG9wM/edit?usp=sharing)).
2. Install the app on an Android device.
3. Verify that the experiment from prerequisites is displayed in the secret menu.
4. End the experiment using the Nimbus console and the help of a colleague.
5. Go to “[https://settings-writer.stage.mozaws.net/v1/buckets/main/collections/nimbus-mobile-experiments/records](https://settings-writer.stage.mozaws.net/v1/buckets/main/collections/nimbus-mobile-experiments/records)” to verify that the change is displayed there.
6. Completely close the app and open it again a few times.
7. Observe the list of experiments displayed in the secret menu.

**[Expected result]:**

* The experiment is no longer displayed after two startups.

**[Actual result]:**

* The experiment remains active until some time has passed, which varies in length. Sometimes the change is picked up after a few minutes, sometimes it takes even 2 hours.
* The number of restarts does not seem to affect the result.

**[Notes]:**

* This applies to launching experiments or ending enrollment as well.

┆Issue is synchronized with this [Jira Bug](https://mozilla-hub.atlassian.net/browse/EXP-2078)
┆Attachments: <a href="https://mozilla-hub.atlassian.net/secure/attachment/19164/image-20220111-123130.png">image-20220111-123130.png</a>

**Comment by data-sync-user** [comment]
➤ James Hugman commented:

I think this a feature of the caching/backoff/retry logic in the SDK. Could you find the Backoff or the Retry-After headers? If you could delete those headers (e.g. using mitmproxy) and still see the bug occurring this would be worrying.

**Comment by data-sync-user** [comment]
➤ James Hugman commented:

Maria Heres could you retest as suggested; otherwise, this looks like it’s working as suggested.

**Comment by data-sync-user** [comment]
➤ Maria Heres commented:

James Hugman I have begun trying the instructions using mitmproxy without much success. I first tried to find the Backoff or the Retry-After headers, but had no success finding them looking at the requests displayed in mitmproxy when opening Firefox Nightly 97 (with the mitmproxy certificate for Android installed as a User certificate). I have not tried to do this before and am a bit unsure how to proceed from this point. Could you please provide more detailed steps?

**Comment by data-sync-user** [comment]
➤ James Hugman commented:

!image-20220111-123130.png|width=859,height=389!

Above are the response headers from https://settings-cdn.stage.mozaws.net/v1/buckets/main/collections/nimbus-mobile-experiments/records ( https://settings-cdn.stage.mozaws.net/v1/buckets/main/collections/nimbus-mobile-experiments/records ) as shown in the JSON View of Firefox. I hadn’t thought to do this with the browser before: this is much simpler than mitm!

I see that max-age is 3600. I’m not sure what the SDK is supposed to be doing with this header, or if the Retry-After or Backoff headers are set when requested with the SDK.

I also see you’re relying on the secret menu to observe changes.

I’m puzzled that if this bug is as reported, then something is very broken and experiments wouldn’t be working for anyone.

Next steps would be to try the steps to reproduce, but wait at least one hour (3600s) between the update/restart dance. If experiments are updating then, then we should close this as INVALID; if they aren’t, then this should be upgraded to P1/S1.

**Comment by data-sync-user** [comment]
➤ James Hugman commented:

Further future context:

  * the https connection is going to be managed by the Remote-Settings rust client. This should provide high quality cache control.
  * viaduct is likely to be uniffi’d, removing protobufs. This should be transparent to the SDK.

**Comment by data-sync-user** [comment]
➤ Maria Heres commented:

For three updates, two unenrollments and one enrollment, using four Android devices, the changes were picked up after two restarts. What we observed is that the changes were picked up at the same as they appeared in https://settings-cdn.stage.mozaws.net/v1/buckets/main/collections/nimbus-mobile-experiments/records. ( https://settings-cdn.stage.mozaws.net/v1/buckets/main/collections/nimbus-mobile-experiments/records. ) The time it took them to appear there varied from 20 minutes to over an hour (change observed at 1.5 hours), but once they were displayed, we were able to see them on the devices as well. As a result, I think this issue can be closed as INVALID.
Tested using Samsung Galaxy S8 - Android 9, Google Pixel 2 XL - Android 10, HUAWEI P20 lite - Android 8.0.0, OnePlus 6 - Android 10 on Firefox Nightly 97.0a1 (Build #2015856835), 622a4f57d+, AC: 97.0.20220109143152, 324faa4260, GV: 97.0a1-20220108220226, AS:87.1.0.
